contracts/acl/ACL.sol

```diff
@@ -0,0 +1,348 @@
+        } else if (param.id == SENDER_PARAM_ID) {
+            value = uint256(msg.sender);
+        } else if (param.id == PARAM_VALUE_PARAM_ID) {
+            value = uint256(param.value);
+        } else {
+            if (param.id >= how.length) {
+                return false;
+            }
+            value = uint256(uint240(how[param.id])); // force lost precision
+        }
+
+        if (Op(param.op) == Op.RET) {
+            return uint256(value) > 0;
+        }
+
+        return compare(value, Op(param.op), comparedTo);
+    }
+
+    function evalLogic(Param param, bytes32 paramsHash, address who, address where, bytes32 what, uint256[] how) internal view returns (bool) {
+        if (Op(param.op) == Op.IF_ELSE) {
+            var (condition, success, failure) = decodeParamsList(uint256(param.value));
+            bool result = evalParam(paramsHash, condition, who, where, what, how);
+
+            return evalParam(paramsHash, result ? success : failure, who, where, what, how);
+        }
+
+        var (v1, v2,) = decodeParamsList(uint256(param.value));
+        bool r1 = evalParam(paramsHash, v1, who, where, what, how);
+
+        if (Op(param.op) == Op.NOT) {
+            return !r1;
+        }
+
+        if (r1 && Op(param.op) == Op.OR) {
+            return true;
+        }
+
+        if (!r1 && Op(param.op) == Op.AND) {
+            return false;
+        }
+
+        bool r2 = evalParam(paramsHash, v2, who, where, what, how);
+
+        if (Op(param.op) == Op.XOR) {
+            return (r1 && !r2) || (!r1 && r2);
+        }
+
+        return r2; // both or and and depend on result of r2 after checks
+    }
+
+    function compare(uint256 a, Op op, uint256 b) internal pure returns (bool) {
+        if (op == Op.EQ)  return a == b;                              // solium-disable-line lbrace
+        if (op == Op.NEQ) return a != b;                              // solium-disable-line lbrace
+        if (op == Op.GT)  return a > b;                               // solium-disable-line lbrace
+        if (op == Op.LT)  return a < b;                               // solium-disable-line lbrace
+        if (op == Op.GTE) return a >= b;                              // solium-disable-line lbrace
+        if (op == Op.LTE) return a <= b;                              // solium-disable-line lbrace
+        return false;
+    }
+
+    /**
+    * @dev Internal function that sets management
+    */
+    function _setPermissionManager(address _newManager, address _app, bytes32 _role) internal {
+        permissionManager[roleHash(_app, _role)] = _newManager;
+        ChangePermissionManager(_app, _role, _newManager);
+    }
+
+    function roleHash(address where, bytes32 what) pure internal returns (bytes32) {
+        return keccak256(uint256(1), where, what);
+    }
+
+    function permissionHash(address who, address where, bytes32 what) pure internal returns (bytes32) {
+        return keccak256(uint256(2), who, where, what);
```

Comment:
What do you think about using a string, e.g. "permission" and "role", to prefix each hash instead?